src/pudl/analysis/allocate_net_gen.py

```diff
@@ -324,183 +509,437 @@ def stack_generators(
     return gens_stack
 
 
-def associate_generator_tables(gf, gen, gens):
-    """Associate the three tables needed to assign net gen to generators.
+def associate_generator_tables(
+    gens: pd.DataFrame, gf: pd.DataFrame, gen: pd.DataFrame, bf_by_gens: pd.DataFrame
+) -> pd.DataFrame:
+    """Associate the three tables needed to assign net gen and fuel to generators.
+
+    The ``generation_fuel_eia923`` table's data is reported at the
+    :py:const:`IDX_PM_ESC` granularity. Each generator has one ``prime_mover_code``, but
+    potentially several ``energy_source_code``s (reported in as primary source,
+    secondary source, etc.). Much of the data in ``generation_fuel_eia923`` not
+    associated with primary energy sources, so before merging the data with the
+    generators, we employ :func:``stack_generators``.
```

Comment:
"**is** not associated with primary energy sources"?

This description doesn't make sense to me. What does it mean for the data in `gf` not to be associated with primary energy sources? What is the data that you're referring to in that table? What does it mean to stack generators?

Reply:
I've explained what primary and secondary fuel sources are right above in the context of generators.

what I mean here is the `gf` table has energy sources as a part of its primary key - and therefor the data columns (net gen and fuel consumed) in the `gf` table as associated with energy sources. But there are many instances in which the `gf` table's `energy_source_code` is not the primary energy source (i.e. the `energy_source_code1`) for any of the generators in that plant.

stack is referring to pandas [stack method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.stack.html)

Reply:
I updated this lmk if you think it is clear enough or if you have suggestions